bitwarden/authenticator-ios
# [BITAU-131] Add Popup to QR Code Scan to Add Code to Bitwarden

## 🎟️ Tracking

[BITAU-131](https://livefront.atlassian.net/browse/BITAU-131)

## 📔 Objective

The main objective of this PR is to add the ability to save QR code scans in the Bitwarden PM app when they originate from the Authenticator app. There's a couple of pieces in support of this:
- A new dialog that appears the first time and asks the user if they would like to save the code locally or send it to Bitwarden
- A second new dialog that appears (only the first time) asking if they'd like to set whatever option they have chosen as their default.
- A new pair of app settings
  - `defaultSaveOption` - this records where the user wants to save codes by default (locally, in Bitwarden, or ask each time)
  - `hasSeenDefaultSaveOptionPrompt` - this tells us if they've seen the dialog asking them to set their default. We only show it once. After that, everything is controlled by the `defaultSaveOption`.
- A new row in Settings to allow the user to change their `defaultSaveOption`.

### Dialog flow
- If the user has seen the dialog about setting their default, we always respect their preference
  - This means we either immediately process the code (locally or sending it to the PM app), or we show the dialog and follow their choice from there.
- If they have  _not_ seen the dialog:
  - First we show them the dialog asking where they'd like to save the code (over the QR scan view)
  - Second, we show them another dialog asking them if they'd like to make that their default (over the Item List view)
  - Finally, we save their choice and take action on their code (either saving locally or sending it to the PM app)

### Testing notes
- As with the other PRs, make sure you have both the PM and Auth feature flags enabled
- Make sure you're using an `otpauth://` style QR code
- Make sure you've already enabled sync for at least one account in the PM app.

## 📸 Screenshots
![IMG_3332](https://github.com/user-attachments/assets/677372b1-f3ef-4314-b7bf-6b62d61b8c6d)
![IMG_3334](https://github.com/user-attachments/assets/61998355-0732-4a7c-8760-b2736ce3ddbd)
![IMG_3333](https://github.com/user-attachments/assets/57ae3796-b475-4e47-a6f0-b75c12e00f9f)
![Simulator Screenshot - iPhone 15 Pro - 2024-10-25 at 12 04 26](https://github.com/user-attachments/assets/cd388e56-1367-46d7-8d13-beb3eec637e2)
![Simulator Screenshot - iPhone 15 Pro - 2024-10-25 at 12 04 24](https://github.com/user-attachments/assets/f504be44-ab5f-4481-bf63-59b9da42dba0)

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines



- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes
